src/library.js

```diff
@@ -967,63 +967,78 @@ addToLibrary({
+      /* month name */      'b': '(?:Jan(?:uary)?)|(?:Feb(?:ruary)?)|(?:Mar(?:ch)?)|(?:Apr(?:il)?)|May|(?:Jun(?:e)?)|(?:Jul(?:y)?)|(?:Aug(?:ust)?)|(?:Sep(?:tember)?)|(?:Oct(?:ober)?)|(?:Nov(?:ember)?)|(?:Dec(?:ember)?)',
+      /* century */         'C': '\\d\\d',
+      /* day of month */    'd': '0[1-9]|[1-9](?!\\d)|1\\d|2\\d|30|31',
+      /* hour (24hr) */     'H': '\\d(?!\\d)|[0,1]\\d|20|21|22|23',
+      /* hour (12hr) */     'I': '\\d(?!\\d)|0\\d|10|11|12',
+      /* day of year */     'j': '00[1-9]|0?[1-9](?!\\d)|0?[1-9]\\d(?!\\d)|[1,2]\\d\\d|3[0-6]\\d',
+      /* month */           'm': '0[1-9]|[1-9](?!\\d)|10|11|12',
+      /* minutes */         'M': '0\\d|\\d(?!\\d)|[1-5]\\d',
+      /* whitespace */      'n': '\\s',
+      /* AM/PM */           'p': 'AM|am|PM|pm|A\\.M\\.|a\\.m\\.|P\\.M\\.|p\\.m\\.',
+      /* seconds */         'S': '0\\d|\\d(?!\\d)|[1-5]\\d|60',
+      /* week number */     'U': '0\\d|\\d(?!\\d)|[1-4]\\d|50|51|52|53',
+      /* week number */     'W': '0\\d|\\d(?!\\d)|[1-4]\\d|50|51|52|53',
+      /* weekday number */  'w': '[0-6]',
+      /* 2-digit year */    'y': '\\d\\d',
+      /* 4-digit year */    'Y': '\\d\\d\\d\\d',
+      /* whitespace */      't': '\\s',
+      /* time zone */       'z': 'Z|(?:[\\+\\-]\\d\\d:?(?:\\d\\d)?)'
     };
 
     var MONTH_NUMBERS = {JAN: 0, FEB: 1, MAR: 2, APR: 3, MAY: 4, JUN: 5, JUL: 6, AUG: 7, SEP: 8, OCT: 9, NOV: 10, DEC: 11};
     var DAY_NUMBERS_SUN_FIRST = {SUN: 0, MON: 1, TUE: 2, WED: 3, THU: 4, FRI: 5, SAT: 6};
     var DAY_NUMBERS_MON_FIRST = {MON: 0, TUE: 1, WED: 2, THU: 3, FRI: 4, SAT: 5, SUN: 6};
 
-    for (var datePattern in DATE_PATTERNS) {
-      pattern = pattern.replace(datePattern, '('+datePattern+DATE_PATTERNS[datePattern]+')');
-    }
-
-    // take care of capturing groups
+    var pattern_out = "";
+    var in_percent = false;
     var capture = [];
-    for (var i=pattern.indexOf('%'); i>=0; i=pattern.indexOf('%')) {
-      capture.push(pattern[i+1]);
-      pattern = pattern.replace(new RegExp('\\%'+pattern[i+1], 'g'), '');
+    // go through each character in the string finding 
+    // the %X %Y etc. patterns. We used to do this with find, but that 
+    // approach caused bugs when doing something like 
+    // %%Z (a literal % followed by a Z). This is less fragile.
+    for (var i = 0; i < pattern.length; i++) {
+      var c = pattern[i];
+      if (c === "%") {
+        if (in_percent) {
+          pattern_out += "%";
+        }
+        in_percent = !in_percent;
+      } else if (in_percent){
+        if (c in EQUIVALENT_MATCHERS){
+          // replace this pattern with the equivalent then
+          // reset the loop to this position
+          pattern=pattern.slice(0,i-1)+EQUIVALENT_MATCHERS[c]+pattern.slice(i+1)
+          i=i-2;
+        }else if (c in DATE_PATTERNS) {
+          capture.push(c)
+          pattern_out += "(" + DATE_PATTERNS[c] + ")";
+        }
+        in_percent = false;
+      } else {
+        pattern_out += c;
```

Comment:
Btw after writing my implementation in that comment I figured there is a bug here - this doesn't add `\` to escape the character, so special symbols like `.` will be preserved in the final pattern and will match arbitrary character instead of just a dot.

Same in the previous implementation which didn't have any handling for those special characters.

Reply:
The special symbols are escaped at the top no?

Reply:
Sigh, you're right. I expected to see that done only for remaining literals after % replacements so didn't look there, and didn't see any tests in the test file ensuring that eg dot doesn't accidentally match unrelated character.